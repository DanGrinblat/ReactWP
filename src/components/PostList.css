.post-list {
	display: grid;
	position: absolute;
	top: 50%;
	grid-template-columns: repeat(auto-fill, minmax(300px, 1fr));
	grid-gap: 0 10px;
	grid-auto-rows: 10px;
	vertical-align: middle;
}

/* Make room for pagination module and footer */
.ui.grid {
	margin-top: 0 !important;
	margin-bottom: -58px !important;
	min-height: calc(100vh - 100px);
}
