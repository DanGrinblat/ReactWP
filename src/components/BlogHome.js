import React from "react";
import PostList from "./PostList";
import Pagination from "./Pagination";
import wprest from "../api/wprest";

class BlogHome extends React.Component {
	state = {
		posts: [],
		media: [],
		categories: [],
		pageCount: 1,
		postCount: 0,
		postsPerPage: 6,
		selectedPage: 1
	};

	componentDidMount() {
		this.retrieveAll(this.state.selectedPage, this.state.postsPerPage);
	}

	onClickPage = newPage => {
		this.setState({
			selectedPage: Number(newPage)
		});
		this.retrieveAll(newPage, this.state.postsPerPage);
	};

	//Retrieve Categories, Posts, and Media in succession
	retrieveAll = async (page, perPage) => {
		const categoriesData = await this.retrieveCategories();
		const postsData = await this.retrievePosts(
			page,
			perPage,
			categoriesData
		);

		const postIdsList = postsData.map(post => {
			return post.featured_media;
		});
		await this.retrieveMedia(postIdsList);
	};

	retrievePosts = async (page, perPage, categories) => {
		var params = { page: page, per_page: perPage };

		const selectedCategory = this.props.match.params.categoryterm;
		if (selectedCategory != null) {
			var selectedCategoryExists = false;
			for (var i = 0; i < categories.length; i++) {
				if (selectedCategory === categories[i].slug) {
					params.categories = [categories[i].id];
					selectedCategoryExists = true;
					break;
				}
			}
			if (!selectedCategoryExists) {
				return [];
			}
		}

		const response = await wprest.get("wp/v2/posts", { params });
		console.log(response);

		if (this.state.posts !== response.data) {
			//If there are 0 pages, set pageCount to 1
			const newPageCount =
				response.headers["x-wp-totalpages"] !== "0"
					? response.headers["x-wp-totalpages"]
					: "1";
			this.setState({
				posts: response.data,
				postCount: response.headers["x-wp-total"],
				pageCount: newPageCount
			});
		}
		return response.data;
	};

	async retrieveMedia(postIdsList) {
		const response = await wprest.get("wp/v2/media", {
			params: { include: postIdsList }
		});

		if (this.state.media !== response.data) {
			this.setState({ media: response.data });
		}
		return response.data;
	}

	async retrieveCategories() {
		const response = await wprest.get("wp/v2/categories");
		if (this.state.categories !== response.data) {
			this.setState({ categories: response.data });
		}
		return response.data;
	}

	render() {
		return (
			<div>
				<PostList posts={this.state.posts} media={this.state.media} />
				<center>
					<Pagination
						selectedPage={this.state.selectedPage}
						pageCount={this.state.pageCount}
						postCount={this.state.postCount}
						onClickPage={this.onClickPage}
					/>
				</center>
			</div>
		);
	}
}

export default BlogHome;
