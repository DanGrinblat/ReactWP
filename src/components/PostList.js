import "./PostList";
import React from "react";
import Post from "./Post";

class PostList extends React.Component {
	renderList() {
		return this.props.posts.map(post => {
			for (var i in this.props.media) {
				if (post.featured_media === this.props.media[i].id) {
					post.imageLink = this.props.media[i].source_url;
				}
			}

			return (
				<div key={post.id} className="equal width column">
					<center>
						<Post post={post} />
					</center>
				</div>
			);
		});
	}

	render() {
		return (
			<div className="ui container three column doubling stackable middle aligned centered grid post-list">
				{this.renderList()}
			</div>
		);
	}
}

export default PostList;
