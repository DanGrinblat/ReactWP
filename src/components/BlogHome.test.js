import React from "react";
import { shallow, mount } from "enzyme";
import { Route } from "react-router";
import BlogHome from "./BlogHome";
import mockAxios from "jest-mock-axios";
import PostList from "./PostList";

jest.mock("axios");

describe("BlogHome", () => {
	const blogHome = shallow(<BlogHome />);

	it("should render correctly", () => {
		expect(blogHome).toMatchSnapshot();
	});

	it("initializes `state`", () => {
		expect(blogHome.state()).toEqual({
			posts: [],
			media: [],
			categories: [],
			pageCount: 1,
			postCount: 0,
			postsPerPage: 6,
			selectedPage: 1
		});
	});

	describe("when clicking a page", () => {
		it("calls retrieve all", () => {
			blogHome.instance().onClickPage(1);
			expect(mockAxios.get).toHaveBeenCalledWith("wp/v2/categories");
		});
	});
});

describe("API requests", () => {
	let blogHome;
	const props = { params: { categoryterm: null } };

	beforeEach(() => {
		mockAxios.reset();
		blogHome = mount(<BlogHome match={props} />);
	});

	it("retrieves posts with axios", async () => {
		blogHome.instance().retrievePosts(1, 6, []);
		let responseObj = { data: [3, 2, 1] };
		mockAxios.mockResponse(responseObj);

		expect(mockAxios.get).toHaveBeenCalledWith("wp/v2/posts", {
			params: {
				page: 1,
				per_page: 6
			}
		});
	});

	it("retrieves media with axios", async () => {
		blogHome.instance().retrieveMedia([1, 2, 3]);
		let responseObj = { data: [3, 2, 1] };
		mockAxios.mockResponse(responseObj);

		expect(mockAxios.get).toHaveBeenCalledWith("wp/v2/media", {
			params: { include: [1, 2, 3] }
		});
	});

	it("retrieves categories with axios", async () => {
		blogHome.instance().retrieveCategories();
		let responseObj = { data: [1, 2, 3] };
		mockAxios.mockResponse(responseObj);

		expect(mockAxios.get).toHaveBeenCalledWith("wp/v2/categories");
	});
});
