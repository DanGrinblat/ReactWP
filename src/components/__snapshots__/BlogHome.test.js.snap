// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlogHome should render correctly 1`] = `
<div>
  <PostList
    media={Array []}
    posts={Array []}
  />
  <center>
    <Pagination
      onClickPage={[Function]}
      pageCount={1}
      postCount={0}
      selectedPage={1}
    />
  </center>
</div>
`;
